Kocherga/profile epic_2-3

Skoch33 opened the pull request:
Эпик: Профиль
Архитектура: MVVM
Верстка кодом

Файл декомпозиции:
https://github.com/Skoch33/iOS-FakeNFT-StarterProject-Public/blob/kocherga/profile-2-3/Doc/Profile.md
Task tracker:
https://github.com/users/Skoch33/projects/2

- Реализация NavigationBar (est: 30 min; fact: 20 min)
- Создание UITableView (est: 60 min; fact: 30 min)
- Верстка кастомной ячейки (est: 50 min; fact: 40 min)
- Реализация функции сортировки (est: 90 min; fact: 40 min)
- Создание модели (est: 20 min; fact: 20 min)
- Создание сетевого слоя (est: 60 min; fact: 40 min)

`Total:` est: 310 min; fact: 190 min

Skoch33 committed:
Экран мои nft, верстка таблицы

Skoch33 committed:
Экран мои nft,  верстка ячейки

Skoch33 committed:
Экрна мои nft, создание сетевого слоя

Skoch33 committed:
Добавлена кнопка сортировки

Skoch33 committed:
Реализация кнопки сортировки

Skoch33 committed:
Реализовал показ заглушки

Skoch33 committed:
Рефакторинг и кодстайл

Skoch33 committed:
Обновил файл декомпозиции

AlekseiTinkov commented on FakeNFT/Profile/UserNFTScreen/Views/NFTCell.swift:
```diff
@@ -0,0 +1,122 @@
+import UIKit
+import Cosmos
```
Классная библиотека. Но сейчас у желтых и серых звезд разный размер. Возможно проблема просто в асетах. Можно обойтись и вовсе без библиотеки, например создав массив из 5ти UIImageView, вставленных в UIStackView.

Skoch33 replied:
Исправил!

AlekseiTinkov commented on FakeNFT/Profile/UserNFTScreen/Views/NFTCell.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    func configure(nft: NFT, authorName: String) {
+        self.nftImageView.kf.setImage(with: URL(string: nft.images[0]))
```
Обрати внимание - в Фигме есть изображения заглушки для не загрузившихся nft

So1ovyov replied:
Обязательно нужна заглушка для случая когда картинки не нашлось

Skoch33 replied:
Исправлено

AlekseiTinkov commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+    // MARK: - Layout methods
+    
+    private func setupViews() {
+        view.backgroundColor = .white
```
Для корректного отображения темной темы здесь стоит указать .nftWhite

Skoch33 replied:
Исправил!

AlekseiTinkov commented on FakeNFT/Profile/UserNFTScreen/UserNFTViewModel.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    func fetchNFT(nftList: [String]) {
+        ProgressHUD.show(NSLocalizedString("ProgressHUD.loading", comment: ""))
```
ViewModel знает подробности UI - сама показывает/скрывает ProgressHUD. В MVVM это не ее обязанность. Здесь можно создать наблюдаемую сущность, например isLoading и в зависимости от ее состояния отображать ProgressHUD во ViewController

So1ovyov replied:
Верно подмечено, проверь свои ViewModel-и в проекте. У них должны быть методы которые дергают ручки во ViewController

AlekseiTinkov commented on FakeNFT/Models/Network/NFTNetworkRequests.swift:
```diff
@@ -0,0 +1,29 @@
+    let nftID: String
+    
+    var endpoint: URL? {
+        URL(string: "https://651ff0d9906e276284c3c20a.mockapi.io/api/v1/nft/\(nftID)")
```
Базовую часть URL лучше вынести в отдельную константу и использовать по всему проекту

So1ovyov replied:
Хорошая идея, например `private enum Constants`

AlekseiTinkov reviewed (changes requested):
Привет! Отличный код! Круто, что сразу ввел протокол для ViewModel, так будет удобнее делать тесты. Так же очень понравилось как ты грузишь информацию о nft одновременно в нескольких потоках, используя DispatchGroup(). К сожалению в коде нарушен принцип MVVM - ViewModel не должен ничего знать об UI, включая ProgressHUD. Предлагаю попробовать исправить это.

Skoch33 committed:
Правки после командного ревью

Skoch33 committed:
Merge remote-tracking branch 'refs/remotes/origin/kocherga/profile-2-3'

So1ovyov commented on FakeNFT/DesignSystem/UIViewController + Extensions.swift:
```diff
@@ -0,0 +1,17 @@
+extension UIViewController {
+
+    func setupCustomBackButton() {
+        let backButton = UIBarButtonItem(image: UIImage(named: "backward"),
```
Кажется избыточным, можно например:
```
    func setupCustomBackButton() {
        navigationItem.leftBarButtonItem = UIBarButtonItem(
            image: UIImage(named: "backward"),
            style: .plain,
            target: self,
            action: #selector(self.customBackAction)
        )
    }
```

Skoch33 replied:
Исправлено

So1ovyov commented on FakeNFT/Helpers/AlertService.swift:
```diff
@@ -1,53 +1,51 @@
-            textField.placeholder = textFieldPlaceholder
+    
+    func showAlert(model: AlertModel) {
+        let alertController = UIAlertController(title: model.title, message: model.message, preferredStyle: model.style)
```
Давай длинные строки дробить, чтобы легче читалось:
```
        let alertController = UIAlertController(
            title: model.title,
            message: model.message,
            preferredStyle: model.style
        )

```

Skoch33 replied:
Исправлено

So1ovyov commented on FakeNFT/Profile/EditScreen/EditingViewController.swift:
```diff
@@ -87,37 +87,56 @@ final class EditingViewController: UIViewController {
-                                             message: nil,
-                                             textFieldPlaceholder: "URL изображения") { [weak self] urlText in
+    @objc
+    private func changePhotoTapped() {
```
Очень большой метод, давай декомпозируем и вынесем какую-то логику.

So1ovyov commented on FakeNFT/Profile/EditScreen/EditingViewModel.swift:
```diff
@@ -0,0 +1,124 @@
+    
+    private let imageValidator: ImageValidatorProtocol
+    
+    init(profileService: ProfileService, imageValidator: ImageValidatorProtocol = ImageValidator()) {
```
Длинная строка:
```
    init(
        profileService: ProfileService,
        imageValidator: ImageValidatorProtocol = ImageValidator()
    ) {
        self.profileService = profileService
        self.imageValidator = imageValidator
    }
```

So1ovyov replied:
Также давай в ините не задавать дефолтное значение, лучше это делать в месте где ты объявляешь `EditingViewModel` в идеале сделать так:
```
 init(
        profileService: ProfileService,
        imageValidator: ImageValidatorProtocol
    ) {
        self.profileService = profileService
        self.imageValidator = imageValidator
    }
```

So1ovyov commented on FakeNFT/Profile/ProfileScreen/ProfileViewModel.swift:
```diff
@@ -2,98 +2,103 @@ import Foundation
+    init(model: ProfileService, imageValidator: ImageValidatorProtocol = ImageValidator()) {
         self.model = model
         self.imageValidator = imageValidator
+        NotificationCenter.default.addObserver(self, selector: #selector(profileUpdated), name: NSNotification.Name("profileUpdated"), object: nil)
```
Длинная конструкция, давай дробить

Skoch33 replied:
Исправлено

So1ovyov commented on FakeNFT/Profile/UserNFTScreen/UserNFTViewModel.swift:
```diff
@@ -0,0 +1,122 @@
+                case .success(let nft):
+                    fetchedNFTs.append(nft)
+                case .failure(let error):
+                    print("Failed to fetch NFT with ID \(element): \(error)")
```
Было бы круто отображать ошибки в алертах пользователю, чтобы он понимал что происходит в приложении. Принт он не увидит к сожалению

So1ovyov commented on FakeNFT/Profile/UserNFTScreen/Views/NFTCell.swift:
```diff
@@ -0,0 +1,136 @@
+    private lazy var starsImage: [UIImageView] = {
+        (1...5).map { _ in
+            let view = UIImageView()
+            view.image = UIImage()
```
Здесь нет никакой картинки, из-за этого звезд рейтинга нет на экране `Мои NFT`

Skoch33 replied:
Исправлено, была ошибка в написании имени

So1ovyov commented on FakeNFT/Profile/UserNFTScreen/Views/NFTCell.swift:
```diff
@@ -0,0 +1,136 @@
+import Kingfisher
+
+final class NFTCell: UITableViewCell, ReuseIdentifying {
+    private let nftImageView = ViewFactory.shared.createNFTImageView()
```
Почему ViewFactory сделан как Singletone ? Вот рассуждение о том почему это плохо [ссылка](https://stackoverflow.com/questions/35500492/how-to-create-a-uiview-singleton-in-swift)

Давай избавляться от Singletone в проекте для создания верстки. Это плохая затея которая может отработать не корректно.

Skoch33 replied:
Спасибо!
Избавился от этого файла

So1ovyov commented on FakeNFT/Profile/UserNFTScreen/Views/UserNFTViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        cell.selectionStyle = .none
+        
+        guard let nft = viewModel.userNFT?[indexPath.row] else {
+            print("error to get NFT")
```
Давай избавимся от print-ов для дебага в проекте

So1ovyov commented on FakeNFT/Profile/ProfileScreen/Views/ProfileViewController.swift:
```diff
@@ -107,7 +106,7 @@ final class ProfileViewController: UIViewController {
+    private func updateUI(with model: UserProfile) {
         DispatchQueue.main.async { [weak self] in
             self?.profileImageView.kf.setImage(with: URL(string: model.avatar)) { result in
                 switch result {
```
Тут ниже код:
```
[self!.editButton, self!.profileImageView, self!.userNameLabel, self!.userDescriptionLabel, self!.userWebSiteTextView, self!.profileTableView].forEach { $0.isHidden = false }
```
force self нельзя использовать
Вообще любой force unwrap/property/cast запрещен

Skoch33 committed:
Исключил force-операции в проекте
Поправил отображение рейтинга в ячейке
Сделал сортировку по цене и по рейтингу от большего значения к меньшему
Избавился от Singletone, удалил ViewFactory
Поставил заглушку для случая когда картинки не нашлось
Почистил код, чтобы легче читалось

Skoch33 committed:
Убрал пустые файлы

Skoch33 committed:
Исправил ошибку отображения рейтинга

Skoch33 committed:
fix багов
